cpp/src/arrow/array/data.cc

```diff
@@ -298,6 +298,7 @@ void FillZeroLengthArray(const DataType* type, ArraySpan* span) {
 void ArraySpan::FillFromScalar(const Scalar& value) {
   static uint8_t kTrueBit = 0x01;
   static uint8_t kFalseBit = 0x00;
+  auto* scratch_space = const_cast<Scalar&>(value).scratch_space_.data();
```

Comment:
Ouch :-(

Reply:
This can be avoided by using a plan array instead of `std::array`.

Reply:
To be clear, we'll still need the const_cast
```suggestion
  auto* scratch_space = const_cast<Scalar&>(value).scratch_space_;
```

Reply:
Even with `scratch_space_` declared as `mutable`?

Reply:
Nope, mutable spares us that. Forgot to remove the const_cast after I added it.